File: apollo/src/main/java/io/muun/apollo/data/preferences/KeysRepository.java
package io.muun.apollo.data.preferences;

import io.muun.apollo.data.os.secure_storage.SecureStorageProvider;
import io.muun.apollo.data.preferences.adapter.PublicKeyPreferenceAdapter;
import io.muun.apollo.data.preferences.rx.Preference;
import io.muun.apollo.domain.errors.BugDetected;
import io.muun.apollo.domain.errors.MissingMigrationError;
import io.muun.common.crypto.ChallengePublicKey;
import io.muun.common.crypto.ChallengeType;
import io.muun.common.crypto.hd.PrivateKey;
import io.muun.common.crypto.hd.PublicKey;
import io.muun.common.crypto.hd.PublicKeyPair;
import io.muun.common.crypto.hd.Schema;
import io.muun.common.rx.ObservableFn;
import io.muun.common.utils.Preconditions;

import android.content.Context;
import org.bitcoinj.core.NetworkParameters;
import org.threeten.bp.ZonedDateTime;
import rx.Observable;
import timber.log.Timber;

import java.util.NoSuchElementException;
import java.util.Objects;
import javax.annotation.Nullable;
import javax.inject.Inject;
import javax.validation.constraints.NotNull;

public class KeysRepository extends BaseRepository {

    private static final int SECONDS_IN_A_DAY = 60 * 60 * 24;

    // Storage-level keys:
    private static final String KEY_BASE_PRIVATE_KEY_PATH = "xpriv_path";
    private static final String KEY_BASE_58_PRIVATE_KEY = "key_base_58_private_key";
    private static final String KEY_BASE_PUBLIC_KEY = "base_public_key";

    private static final String KEY_CHALLENGE_PUBLIC_KEY = "key_challenge_public_key_";

    private static final String KEY_ENCRYPTED_PRIVATE_USER_KEY = "key_encrypted_private_apollo_key";
    private static final String KEY_ENCRYPTED_PRIVATE_MUUN_KEY = "key_encrypted_private_muun_key";

    private static final String KEY_BASE_MUUN_PUBLIC_KEY = "base_muun_public_key";

    private static final String KEY_MAX_USED_EXTERNAL_ADDRESS_INDEX =
            "key_max_used_external_address_index";

    private static final String KEY_MAX_WATCHING_EXTERNAL_ADDRESS_INDEX =
            "key_max_watching_external_address_index";

    private static final String KEY_EK_ACTIVATION_CODE = "ek_activation_code";

    // Reactive preferences:
    private final Preference<String> basePrivateKeyPathPreference;

    private final Preference<PublicKey> basePublicKeyPreference;
    private final Preference<PublicKey> baseMuunPublicKeyPreference;

    private final Preference<Integer> maxUsedExternalAddressIndexPreference;
    private final Preference<Integer> maxWatchingExternalAddressIndexPreference;

    private final Preference<String> emergencyKitVerificationCode;

    // Dependencies:
    private final NetworkParameters networkParameters;
    private final SecureStorageProvider secureStorageProvider;
    private final UserRepository userRepository;

    /**
     * Creates a KeysRepository.
     */
    @Inject
    public KeysRepository(
            Context context,
            NetworkParameters networkParameters,
            SecureStorageProvider secureStorageProvider,
            UserRepository userRepository) {

        super(context);

        this.networkParameters = networkParameters;

        this.secureStorageProvider = secureStorageProvider;

        this.userRepository = userRepository;

        basePrivateKeyPathPreference = rxSharedPreferences
                .getString(KEY_BASE_PRIVATE_KEY_PATH);

        maxUsedExternalAddressIndexPreference = rxSharedPreferences
                .getInteger(KEY_MAX_USED_EXTERNAL_ADDRESS_INDEX);

        maxWatchingExternalAddressIndexPreference = rxSharedPreferences
                .getInteger(KEY_MAX_WATCHING_EXTERNAL_ADDRESS_INDEX);

        basePublicKeyPreference = rxSharedPreferences.getObject(
                KEY_BASE_PUBLIC_KEY,
                PublicKeyPreferenceAdapter.INSTANCE
        );

        baseMuunPublicKeyPreference = rxSharedPreferences.getObject(
                KEY_BASE_MUUN_PUBLIC_KEY,
                PublicKeyPreferenceAdapter.INSTANCE
        );

        emergencyKitVerificationCode = rxSharedPreferences
                .getString(KEY_EK_ACTIVATION_CODE);
    }

    /**
     * Returns a new non-stored root PrivateKey.
     */
    @NotNull
    public PrivateKey createRootPrivateKey() {
        return PrivateKey.getNewRootPrivateKey(networkParameters);
    }

    /**
     * Returns the current base private key.
     */
    @NotNull
    public Observable<PrivateKey> getBasePrivateKey() {
        return secureStorageProvider.getAsync(KEY_BASE_58_PRIVATE_KEY)
                .map(base58Bytes -> {
                    final String path = basePrivateKeyPathPreference.get();
                    final String base58 = new String(base58Bytes);

                    return PrivateKey.deserializeFromBase58(path, base58);
                });
    }

    /**
     * Returns the base public key.
     */
    public PublicKey getBasePublicKey() {
        final PublicKey publicKey = basePublicKeyPreference.get();
        Preconditions.checkState(publicKey != null, "PublicKey must NOT be null!");

        return publicKey;
    }

    /**
     * Returns the base Muun cosigning public key.
     */
    public PublicKey getBaseMuunPublicKey() {
        final PublicKey publicKey = baseMuunPublicKeyPreference.get();
        Preconditions.checkState(publicKey != null, "Muun cosigning PublicKey must NOT be null!");

        return publicKey;
    }

    /**
     * Returns the base user-Muun public key pair.
     */
    public PublicKeyPair getBasePublicKeyPair() {
        return new PublicKeyPair(
                getBasePublicKey(),
                getBaseMuunPublicKey()
        );
    }

    /**
     * Save the base Muun cosigning public key.
     */
    public void storeBaseMuunPublicKey(@NotNull PublicKey publicKey) {
        baseMuunPublicKeyPreference.set(publicKey);
    }

    /**
     * Saves the base private key from which all are derived.
     */
    public Observable<Void> storeBasePrivateKey(@NotNull PrivateKey privateKey) {
        // The private key can be rooted at different derivation paths depending on
        // when it was generated. Older apps rooted them at m. The now expected path is
        // Schema#getBasePath, so we derive to it to ensure consistency for older keys.
        final PrivateKey basePrivateKey = privateKey.deriveFromAbsolutePath(Schema.getBasePath());

        final byte[] base58 = basePrivateKey.serializeBase58().getBytes();

        return secureStorageProvider.putAsync(KEY_BASE_58_PRIVATE_KEY, base58)
                .flatMap(ignored -> storeEncryptedBasePrivateKey(basePrivateKey))
                .doOnNext(ignored -> {
                    basePrivateKeyPathPreference.set(basePrivateKey.getAbsoluteDerivationPath());

                    basePublicKeyPreference.set(basePrivateKey.getPublicKey());
                    maxUsedExternalAddressIndexPreference.set(-1); // set a default
                });
    }

    private Observable<Void> storeEncryptedBasePrivateKey(@NotNull PrivateKey basePrivateKey) {
        // We want to keep the flow on an Observable<View> which has `doOnNext` transformations.
        // In order to do that we are always returning a result, so that onNext is called on
        // the observable. `onErrorResumeNext(Observable.just(null))` keeps the flow on errors
        // and the `return Observable.just(null)` on `flatMap` is just passing that that null.

        if (!hasChallengePublicKey(ChallengeType.RECOVERY_CODE)) {
            // Without challenge keys, we can't store the exportable encrypted private key:
            return Observable.just(null);
        }

        return getChallengePublicKey(ChallengeType.RECOVERY_CODE)
                .map(challengePublicKey -> {
                    final long birthday = getWalletBirthdaySinceGenesis();
                    return challengePublicKey.encryptPrivateKey(basePrivateKey, birthday);
                })
                .onErrorResumeNext(error -> {
                    Timber.e(error);
                    return Observable.just(null);
                })
                .flatMap(encryptedKey -> {
                    if (encryptedKey == null) {
                        return Observable.just(null);
                    }

                    Timber.d("Storing encrypted Apollo private key in secure storage.");

                    return secureStorageProvider.putAsync(
                            KEY_ENCRYPTED_PRIVATE_USER_KEY,
                            encryptedKey.getBytes()
                    );
                });
    }

    public void storeEmergencyKitVerificationCode(String verificationCode) {
        emergencyKitVerificationCode.set(verificationCode);
    }

    public Observable<String> watchEmergencyKitVerificationCode() {
        return emergencyKitVerificationCode.asObservable();
    }

    private long getWalletBirthdaySinceGenesis() {

        final ZonedDateTime createdAt = userRepository.fetchOneOptional()
                .flatMap(it -> it.createdAt)
                .orElse(null);

        if (createdAt == null) {
            return 0xFFFF; // a placeholder, only for the user key. Not ideal, not necessary.
        }

        final long creationTimestamp = createdAt.toEpochSecond();
        final long genesisTimestamp = networkParameters.getGenesisBlock().getTimeSeconds();
        return (creationTimestamp - genesisTimestamp) / SECONDS_IN_A_DAY;
    }

    public Observable<String> getEncryptedBasePrivateKey() {
        return secureStorageProvider.getAsync(KEY_ENCRYPTED_PRIVATE_USER_KEY)
                .map(String::new);
    }

    /**
     * Obtain from secure storage the ChallengePublicKey for a ChallengeType.
     */
    public Observable<ChallengePublicKey> getChallengePublicKey(ChallengeType type) {

        return secureStorageProvider
                .getAsync(KEY_CHALLENGE_PUBLIC_KEY + type.toString())
                .map(serialization -> {

                    // Latest ChallengePublicKey.deserialize() assumes pub key serialization has
                    // salt (and pub key version). If the length of the pub key serialization
                    // doesn't include the salt, it means the salt migration hasn't run. We need
                    // that to run to properly migrated ChallengePublicKey to have salt and version.
                    if (serialization.length == ChallengePublicKey.PUBLIC_KEY_LENGTH) {
                        throw new MissingMigrationError("ChallengePublicKey salt");
                    }

                    final ChallengePublicKey challengePublicKey = ChallengePublicKey.deserialize(
                            serialization
                    );

                    // NOTE:
                    // Old Apollo versions did not store the challenge key salt. Modern Apollo
                    // versions concatenate the salt to the public key bytes (see the store method
                    // below). We use the length of the stored key to determine whether the salt
                    // is present (the subArray will be empty if not):
                    if (Objects.requireNonNull(challengePublicKey.getSalt()).length == 0) {
                        throw new MissingMigrationError("ChallengePublicKey salt");
                    }

                    return challengePublicKey;
                });
    }

    /**
     * Return true if ChallengeKeys have already been migrated.
     */
    public boolean hasMigratedChallengeKeys() {
        return getChallengePublicKey(ChallengeType.PASSWORD)
                .map(key -> true)
                .compose(ObservableFn.onTypedErrorResumeNext(
                        NoSuchElementException.class,
                        error -> Observable.just(true) // nothing to migrate if the user is UU
                ))
                .compose(ObservableFn.onTypedErrorResumeNext(
                        MissingMigrationError.class,
                        error -> Observable.just(false)
                ))
                .toBlocking()
                .first();
    }

    /**
     * Return true if the public key for a ChallengeType is present in storage.
     */
    public boolean hasChallengePublicKey(ChallengeType type) {
        return secureStorageProvider.has(KEY_CHALLENGE_PUBLIC_KEY + type.toString());
    }

    /**
     * Saves a string containing an encrypted private muun key.
     *
     * @param encryptedPrivateMuunKey the encrypted key plus metadata.
     */
    public void storeEncryptedMuunPrivateKey(String encryptedPrivateMuunKey) {

        Timber.d("Stored encrypted muun key on secure storage.");

        secureStorageProvider.put(
                KEY_ENCRYPTED_PRIVATE_MUUN_KEY,
                encryptedPrivateMuunKey.getBytes()
        );
    }

    public Observable<String> getEncryptedMuunPrivateKey() {
        return secureStorageProvider.getAsync(KEY_ENCRYPTED_PRIVATE_MUUN_KEY)
                .map(String::new);
    }

    /**
     * Stores a public challenge key on secure storage.
     *
     * @param publicKey challenge public key.
     * @param type      the challenge type.
     */
    public void storePublicChallengeKey(ChallengePublicKey publicKey, ChallengeType type) {

        Timber.d("Storing challenge public key %s on secure storage.", type.toString());

        if (publicKey.getSalt() == null) {
            throw new MissingMigrationError("ChallengePublicKey salt");
        }

        secureStorageProvider.put(
                KEY_CHALLENGE_PUBLIC_KEY + type.toString(),
                publicKey.serialize()
        );

        // Update the encrypted Apollo private key when some challenge key changes.
        if (secureStorageProvider.has(KEY_BASE_58_PRIVATE_KEY)) {
            getBasePrivateKey()
                    .flatMap(this::storeEncryptedBasePrivateKey)
                    .toBlocking()
                    .first();
        }
    }

    /**
     * Add the salt to an existing challenge key for old sessions.
     *
     * <p>This is compat method and should not be used for anything else.</p>
     */
    public void seasonPublicChallengeKey(final byte[] salt, final ChallengeType type) {
        Timber.d("Adding salt to %s", type.toString());

        final byte[] publicKey = secureStorageProvider
                .get(KEY_CHALLENGE_PUBLIC_KEY + type.toString());

        if (publicKey.length != ChallengePublicKey.PUBLIC_KEY_LENGTH) {
            Timber.e(new BugDetected("Tried to double-migrate salt for key " + type.toString()));
            return;
        }

        storePublicChallengeKey(ChallengePublicKey.buildLegacy(publicKey, salt), type);
    }

    /**
     * Add version to an existing challenge key. Assumes that it has salt (e.g was created with one
     * or was correctly migrated, see method above).
     *
     * <p>This is compat method and should not be used for anything else.</p>
     */
    public void addChallengePublicKeyVersionMigration(final ChallengeType type) {

        final byte[] legacyPublicKeySerialization = secureStorageProvider
                .get(KEY_CHALLENGE_PUBLIC_KEY + type.toString());

        // If serialization doesn't have salt, it means the salt migration has not run yet. Skip
        // this for now and let salt migration take care of it, when it runs.
        if (legacyPublicKeySerialization.length != ChallengePublicKey.PUBLIC_KEY_LENGTH + 8) {
            return;
        }

        final ChallengePublicKey challengePublicKey = ChallengePublicKey.deserializeLegacy(
                legacyPublicKeySerialization
        );

        storePublicChallengeKey(challengePublicKey, type);
    }

    @Override
    protected String getFileName() {
        return "keys";
    }

    /**
     * Returns the max used external address index or null if none has been used.
     */
    @Nullable
    public Integer getMaxUsedExternalAddressIndex() {
        if (maxUsedExternalAddressIndexPreference.isSet()) {
            return maxUsedExternalAddressIndexPreference.get();
        } else {
            return null;
        }
    }

    public Integer getMaxWatchingExternalAddressIndex() {
        return maxWatchingExternalAddressIndexPreference.get();
    }

    public void setMaxUsedExternalAddressIndex(@Nullable Integer maxUsedExternalAddressIndex) {
        this.maxUsedExternalAddressIndexPreference.set(maxUsedExternalAddressIndex);
    }

    public void setMaxWatchingExternalAddressIndex(Integer maxWatchingExternalAddressIndex) {
        this.maxWatchingExternalAddressIndexPreference.set(maxWatchingExternalAddressIndex);
    }
}

File: android/apollo/src/main/java/io/muun/apollo/data/preferences/KeysRepository.kt
package io.muun.apollo.data.preferences

import android.content.Context
import io.muun.apollo.data.os.secure_storage.SecureStorageProvider
import io.muun.apollo.data.os.secure_storage.SecureStorageProvider.SecureStorageNoSuchElementError
import io.muun.apollo.data.preferences.adapter.PublicKeyPreferenceAdapter
import io.muun.apollo.data.preferences.rx.Preference
import io.muun.apollo.domain.errors.BugDetected
import io.muun.apollo.domain.errors.MissingMigrationError
import io.muun.apollo.domain.utils.toVoid
import io.muun.common.crypto.ChallengePublicKey
import io.muun.common.crypto.ChallengeType
import io.muun.common.crypto.hd.PrivateKey
import io.muun.common.crypto.hd.PublicKey
import io.muun.common.crypto.hd.PublicKeyPair
import io.muun.common.crypto.hd.PublicKeyTriple
import io.muun.common.crypto.hd.Schema
import io.muun.common.rx.ObservableFn
import io.muun.common.utils.Encodings
import io.muun.common.utils.Preconditions
import org.bitcoinj.core.NetworkParameters
import rx.Observable
import timber.log.Timber
import javax.inject.Inject

// Open for mockito to mock/spy
open class KeysRepository @Inject constructor(
    context: Context,
    repositoryRegistry: RepositoryRegistry,
    // Dependencies:
    private val networkParameters: NetworkParameters,
    private val secureStorageProvider: SecureStorageProvider,
    private val userRepository: UserRepository,
) : BaseRepository(context, repositoryRegistry) {

    companion object {
        private const val SECONDS_IN_A_DAY = 60 * 60 * 24

        // Storage-level keys:
        private const val KEY_BASE_PRIVATE_KEY_PATH = "xpriv_path"
        private const val KEY_BASE_58_PRIVATE_KEY = "key_base_58_private_key"
        private const val KEY_BASE_PUBLIC_KEY = "base_public_key"
        private const val KEY_CHALLENGE_PUBLIC_KEY = "key_challenge_public_key_"
        private const val KEY_ENCRYPTED_PRIVATE_USER_KEY = "key_encrypted_private_apollo_key"
        private const val KEY_ENCRYPTED_PRIVATE_MUUN_KEY = "key_encrypted_private_muun_key"
        private const val KEY_MUUN_KEY_FINGERPRINT = "fingerprint_muun_key"
        private const val KEY_USER_KEY_FINGERPRINT = "fingerprint_user_key"
        private const val KEY_BASE_MUUN_PUBLIC_KEY = "base_muun_public_key"
        private const val KEY_BASE_SWAP_SERVER_PUBLIC_KEY = "base_swap_server_public_key"
        private const val KEY_MAX_USED_EXTERNAL_ADDRESS_INDEX =
            "key_max_used_external_address_index"
        private const val KEY_MAX_WATCHING_EXTERNAL_ADDRESS_INDEX =
            "key_max_watching_external_address_index"
    }


    // Reactive preferences:
    private val basePrivateKeyPathPreference: Preference<String>
        get() = rxSharedPreferences.getString(KEY_BASE_PRIVATE_KEY_PATH)
    private val basePublicKeyPreference: Preference<PublicKey>
        get() = rxSharedPreferences.getObject(KEY_BASE_PUBLIC_KEY, PublicKeyPreferenceAdapter.INSTANCE)
    private val baseMuunPublicKeyPreference: Preference<PublicKey>
        get() = rxSharedPreferences.getObject(KEY_BASE_MUUN_PUBLIC_KEY, PublicKeyPreferenceAdapter.INSTANCE)
    private val baseSwapServerPublicKeyPreference: Preference<PublicKey>
        get() = rxSharedPreferences.getObject(
            KEY_BASE_SWAP_SERVER_PUBLIC_KEY,
            PublicKeyPreferenceAdapter.INSTANCE
        )
    private val muunKeyFingerprintPreference: Preference<String>
        get() = rxSharedPreferences.getString(KEY_MUUN_KEY_FINGERPRINT)
    private val userKeyFingerprintPreference: Preference<String>
        get() = rxSharedPreferences.getString(KEY_USER_KEY_FINGERPRINT)
    private val maxUsedExternalAddressIndexPreference: Preference<Int>
        get() = rxSharedPreferences.getInteger(KEY_MAX_USED_EXTERNAL_ADDRESS_INDEX)
    private val maxWatchingExternalAddressIndexPreference: Preference<Int>
        get() = rxSharedPreferences.getInteger(KEY_MAX_WATCHING_EXTERNAL_ADDRESS_INDEX)

    override val fileName get() = "keys"

    /**
     * Returns a new non-stored root PrivateKey.
     */
    fun createRootPrivateKey(): PrivateKey {
        return PrivateKey.getNewRootPrivateKey(networkParameters)
    }

    /**
     * Returns the current base private key.
     */
    val basePrivateKey: Observable<PrivateKey>
        get() = secureStorageProvider.getAsync(KEY_BASE_58_PRIVATE_KEY)
            .map { base58Bytes: ByteArray? ->
                val path = basePrivateKeyPathPreference.get()
                val base58 = String(base58Bytes!!)
                PrivateKey.deserializeFromBase58(path, base58)
            }

    /**
     * Returns the base public key.
     */
    val basePublicKey: PublicKey
        get() {
            val publicKey = basePublicKeyPreference.get()
            Preconditions.checkState(publicKey != null, "PublicKey must NOT be null!")
            return publicKey!!
        }

    /**
     * Returns the base Muun cosigning public key.
     */
    val baseMuunPublicKey: PublicKey
        get() {
            val publicKey = baseMuunPublicKeyPreference.get()
            Preconditions.checkState(
                publicKey != null,
                "Muun cosigning PublicKey must NOT be null!"
            )
            return publicKey!!
        }

    /**
     * Returns the swap server public key for this user.
     */
    private val baseSwapServerPublicKey: PublicKey?
        get() = baseSwapServerPublicKeyPreference.get()

    /**
     * Returns the base user-Muun public key pair.
     */
    @get:Deprecated("use getBasePublicKeyTriple", ReplaceWith("basePublicKeyTriple"))
    val basePublicKeyPair: PublicKeyPair
        get() = PublicKeyPair(
            basePublicKey,
            baseMuunPublicKey
        )

    /**
     * Returns the base user-Muun-swapServer public key triple.
     */
    val basePublicKeyTriple: PublicKeyTriple
        get() = PublicKeyTriple(
            basePublicKey,
            baseMuunPublicKey,
            baseSwapServerPublicKey
        )

    /**
     * Save the base Muun cosigning public key.
     */
    fun storeBaseMuunPublicKey(publicKey: PublicKey) {
        baseMuunPublicKeyPreference.set(publicKey)
    }

    /**
     * Save the base swap server key for this user.
     */
    fun storeSwapServerPublicKey(publicKey: PublicKey) {
        baseSwapServerPublicKeyPreference.set(publicKey)
    }

    /**
     * Saves the base private key from which all are derived.
     */
    fun storeBasePrivateKey(privateKey: PrivateKey): Observable<Void> {
        // The private key can be rooted at different derivation paths depending on
        // when it was generated. Older apps rooted them at m. The now expected path is
        // Schema#getBasePath, so we derive to it to ensure consistency for older keys.
        val basePrivateKey = privateKey.deriveFromAbsolutePath(Schema.getBasePath())
        val base58 = basePrivateKey.serializeBase58().toByteArray()
        return secureStorageProvider.putAsync(KEY_BASE_58_PRIVATE_KEY, base58)
            .flatMap { storeEncryptedBasePrivateKey(basePrivateKey) }
            .doOnNext {
                basePrivateKeyPathPreference.set(basePrivateKey.absoluteDerivationPath)
                basePublicKeyPreference.set(basePrivateKey.publicKey)
                maxUsedExternalAddressIndexPreference.set(-1) // set a default
            }
            .toVoid() // Needed because JAVA!!! (rm after secureStorageProvider is kotlinized)
    }

    private fun storeEncryptedBasePrivateKey(basePrivateKey: PrivateKey): Observable<Void> {
        return if (!hasChallengePublicKey(ChallengeType.RECOVERY_CODE)) {
            // Without challenge keys, we can't store the exportable encrypted private key:
            Observable.just(null)
        } else getChallengePublicKey(ChallengeType.RECOVERY_CODE)
            .map { challengePublicKey: ChallengePublicKey ->
                val birthday = walletBirthdaySinceGenesis
                challengePublicKey.encryptPrivateKey(basePrivateKey, birthday)
            }
            .flatMap { encryptedKey: String ->
                Preconditions.checkNotNull(encryptedKey)
                Timber.d("Storing encrypted Apollo private key in secure storage.")
                storeUserKeyFingerprint(Encodings.bytesToHex(basePrivateKey.fingerprint))
                secureStorageProvider.putAsync(
                    KEY_ENCRYPTED_PRIVATE_USER_KEY,
                    encryptedKey.toByteArray()
                )
            }
    }

    // a placeholder, only for the user key. Not ideal, not necessary.
    private val walletBirthdaySinceGenesis: Long
        get() {
            val createdAt = userRepository.fetchOneOptional()
                .flatMap { it.createdAt }
                .orElse(null)
                ?: return 0xFFFF // a placeholder, only for the user key. Not ideal, not necessary.
            val creationTimestamp = createdAt.toEpochSecond()
            val genesisTimestamp = networkParameters.genesisBlock.timeSeconds
            return (creationTimestamp - genesisTimestamp) / SECONDS_IN_A_DAY
        }

    val encryptedBasePrivateKey: Observable<String>
        get() = secureStorageProvider.getAsync(KEY_ENCRYPTED_PRIVATE_USER_KEY)
            .map { bytes: ByteArray -> String(bytes) }

    /**
     * Obtain from secure storage the ChallengePublicKey for a ChallengeType.
     */
    fun getChallengePublicKey(type: ChallengeType): Observable<ChallengePublicKey> {
        return secureStorageProvider
            .getAsync(KEY_CHALLENGE_PUBLIC_KEY + type.toString())
            .map { serialization: ByteArray ->

                // Latest ChallengePublicKey.deserialize() assumes pub key serialization has
                // salt (and pub key version). If the length of the pub key serialization
                // doesn't include the salt, it means the salt migration hasn't run. We need
                // that to run to properly migrated ChallengePublicKey to have salt and version.
                if (serialization.size == ChallengePublicKey.PUBLIC_KEY_LENGTH) {
                    throw MissingMigrationError("ChallengePublicKey salt")
                }
                val challengePublicKey = ChallengePublicKey.deserialize(serialization)

                // NOTE:
                // Old Apollo versions did not store the challenge key salt. Modern Apollo
                // versions concatenate the salt to the public key bytes (see the store method
                // below). We use the length of the stored key to determine whether the salt
                // is present (the subArray will be empty if not):
                if (requireNotNull(challengePublicKey.salt).isEmpty()) {
                    throw MissingMigrationError("ChallengePublicKey salt")
                }
                challengePublicKey
            }
    }

    /**
     * Return true if ChallengeKeys have already been migrated.
     */
    fun hasMigratedChallengeKeys(): Boolean {
        return getChallengePublicKey(ChallengeType.PASSWORD)
            .map { true }
            .compose(ObservableFn.onTypedErrorResumeNext(
                SecureStorageNoSuchElementError::class.java
            ) { Observable.just(true) } // nothing to migrate if the user is UU
            )
            .compose(ObservableFn.onTypedErrorResumeNext(
                MissingMigrationError::class.java
            ) { Observable.just(false) })
            .toBlocking()
            .first()
    }

    /**
     * Return true if the public key for a ChallengeType is present in storage.
     */
    fun hasChallengePublicKey(type: ChallengeType): Boolean {
        return secureStorageProvider.has(KEY_CHALLENGE_PUBLIC_KEY + type.toString())
    }

    /**
     * Saves a string containing an encrypted private muun key.
     *
     * @param encryptedPrivateMuunKey the encrypted key plus metadata.
     */
    fun storeEncryptedMuunPrivateKey(encryptedPrivateMuunKey: String) {
        Timber.d("Stored encrypted muun key on secure storage.")
        secureStorageProvider.put(
            KEY_ENCRYPTED_PRIVATE_MUUN_KEY,
            encryptedPrivateMuunKey.toByteArray()
        )
    }

    val encryptedMuunPrivateKey: Observable<String>
        get() = secureStorageProvider.getAsync(KEY_ENCRYPTED_PRIVATE_MUUN_KEY)
            .map { bytes: ByteArray -> String(bytes) }

    /**
     * Save a fingerprint corresponding to Muun's private key.
     */
    fun storeMuunKeyFingerprint(muunKeyFingerprint: String) {
        muunKeyFingerprintPreference.set(muunKeyFingerprint)
    }

    val muunKeyFingerprint: Observable<String>
        get() = muunKeyFingerprintPreference.asObservable()

    /**
     * Save a fingerprint corresponding to the user private key.
     */
    fun storeUserKeyFingerprint(userKeyFingerprint: String) {
        userKeyFingerprintPreference.set(userKeyFingerprint)
    }

    val userKeyFingerprint: Observable<String>
        get() = userKeyFingerprintPreference.asObservable()

    /**
     * Stores a public challenge key on secure storage.
     *
     * @param publicKey challenge public key.
     * @param type      the challenge type.
     */
    fun storePublicChallengeKey(publicKey: ChallengePublicKey, type: ChallengeType) {
        Timber.d("Storing challenge public key %s on secure storage.", type.toString())
        if (publicKey.salt == null) {
            throw MissingMigrationError("ChallengePublicKey salt")
        }
        secureStorageProvider.put(
            KEY_CHALLENGE_PUBLIC_KEY + type,
            publicKey.serialize()
        )

        // Update the encrypted Apollo private key when some challenge key changes.
        if (secureStorageProvider.has(KEY_BASE_58_PRIVATE_KEY)) {
            basePrivateKey
                .flatMap { basePrivateKey: PrivateKey ->
                    storeEncryptedBasePrivateKey(basePrivateKey)
                }
                .toBlocking()
                .first()
        }
    }

    /**
     * Add the salt to an existing challenge key for old sessions.
     *
     *
     * This is compat method and should not be used for anything else.
     */
    fun seasonPublicChallengeKey(salt: ByteArray, type: ChallengeType) {
        Timber.d("Adding salt to %s", type.toString())
        val publicKey = secureStorageProvider[KEY_CHALLENGE_PUBLIC_KEY + type.toString()]
        if (publicKey.size != ChallengePublicKey.PUBLIC_KEY_LENGTH) {
            Timber.e(BugDetected("Tried to double-migrate salt for key $type"))
            return
        }
        storePublicChallengeKey(ChallengePublicKey.buildLegacy(publicKey, salt), type)
    }

    /**
     * Add version to an existing challenge key. Assumes that it has salt (e.g was created with one
     * or was correctly migrated, see method above).
     *
     *
     * This is compat method and should not be used for anything else.
     */
    fun addChallengePublicKeyVersionMigration(type: ChallengeType) {
        val legacyPublicKeySerialization =
            secureStorageProvider[KEY_CHALLENGE_PUBLIC_KEY + type.toString()]

        // If serialization doesn't have salt, it means the salt migration has not run yet. Skip
        // this for now and let salt migration take care of it, when it runs.
        if (legacyPublicKeySerialization.size != ChallengePublicKey.PUBLIC_KEY_LENGTH + 8) {
            return
        }
        val challengePublicKey = ChallengePublicKey.deserializeLegacy(
            legacyPublicKeySerialization
        )
        storePublicChallengeKey(challengePublicKey, type)
    }

    /**
     * Returns the max used external address index or null if none has been used.
     */
    var maxUsedExternalAddressIndex: Int?
        get() = if (maxUsedExternalAddressIndexPreference.isSet) {
            maxUsedExternalAddressIndexPreference.get()
        } else {
            null
        }
        /**
         * Save a new max used external address index. This means the maximum index which we have
         * used to generate an address.
         */
        set(maxUsedExternalAddressIndex) {
            maxUsedExternalAddressIndexPreference.set(maxUsedExternalAddressIndex)
        }

    /**
     * Save a new max watching external address index. This means the maximum index which we will
     * using for scanning address for incoming transactions.
     */
    var maxWatchingExternalAddressIndex: Int
        get() = maxWatchingExternalAddressIndexPreference.get()!!
        set(maxWatchingExternalAddressIndex) {
            maxWatchingExternalAddressIndexPreference.set(maxWatchingExternalAddressIndex)
        }
}